_posts/2024-12-02-safety-of-methods-for-numeric-primitive-types.md

```diff
@@ -0,0 +1,195 @@
+
+The same logic applies to the postconditions (`#[ensure]`) of a function.
+
+### 2. Generating Verification Harnesses
+We created macros to generate verification harnesses for each method and type. For example, to generate harnesses for ```unchecked_add```:
+```rust
+#[cfg(kani)]
+mod verify {
+    use super::*;
+
+    macro_rules! generate_unchecked_math_harness {
+        ($type:ty, $method:ident, $harness_name:ident) => {
+            #[kani::proof_for_contract($type::$method)]
+            pub fn $harness_name() {
+                let num1: $type = kani::any::<$type>();
+                let num2: $type = kani::any::<$type>();
+
+                unsafe {
+                    num1.$method(num2);
+                }
+            }
+        }
+    }
+
+    generate_unchecked_math_harness!(i8, unchecked_add, unchecked_add_i8);
+    generate_unchecked_math_harness!(i16, unchecked_add, unchecked_add_i16);
+    // ... Repeat for other integer types
+}
+```
+The harness contains several parts:
+- Symbolic values: These values (```num1``` and ```num2```) are generated by Kani using ```kani::any```.
+- Assumptions: With `#[kani::proof_for_contract]` annotation, Kani automatically inserts `kani::assume()` before the unsafe function call. It ensures that all generated values respect the preconditions of `unchecked_add`. For further details, you can refer to [Kani official RFC for function contracts]([url](https://github.com/model-checking/kani/blob/main/rfc/src/rfcs/0009-function-contracts.md#user-experience)).
+- Unsafe Execution: The invocation of ```unchecked_add``` within an unsafe block for verification.
+
+### 3. Handling Large Input Spaces
+For methods like ```unchecked_mul```, verifying over the entire input space is infeasible due to the exponential number of possibilities. We addressed this by partitioning the input space into intervals:
+```rust
+generate_unchecked_mul_intervals!(i32, unchecked_mul,
+    unchecked_mul_i32_small, -10i32, 10i32,
+    unchecked_mul_i32_large_pos, i32::MAX - 1000i32, i32::MAX,
+    unchecked_mul_i32_large_neg, i32::MIN, i32::MIN + 1000i32,
+    unchecked_mul_i32_edge_pos, i32::MAX / 2, i32::MAX,
+    unchecked_mul_i32_edge_neg, i32::MIN, i32::MIN / 2
+);
+```
+By focusing on critical ranges, we ensured that the verification process remained tractable while still covering important cases where overflows are likely to occur.
+
+The critical ranges include :
+- Small Ranges Near Zero: Includes values around 0, where behavior transitions (e.g., sign changes) are more likely to expose issues like underflows.
+- Boundary Checks: Covers the maximum and minimum values `(i32::MAX, i32::MIN)`, ensuring the method handles edge cases correctly.
```

Comment:
```suggestion
- Boundary Checks: Covers the maximum and minimum values (`i32::MAX`, `i32::MIN`), ensuring the method handles edge cases correctly.
```
to avoid the impression that you are talking about an (open) interval.